### DISCUSSION TITLE
About the hamiltonian evolution

### DISCUSSION THREAD

**Discussion by Yuanningss** [QuSpin/QuSpin#525] [General]
Hi 

I have a question regarding the hamiltonian evolution. 
To calculate state at any time t_f , we have:  H.evolve ( psi_0 , t_0 , t_f). However, this does not specify the steps of the evolution. 
For example, if I want to calculate the state at time t, but with 1000 steps in between, how should I do ? 

thanks !
<img width="715" alt="quspin" src="https://user-images.githubusercontent.com/104153268/164490976-586940a6-9413-4cce-9eaf-b9e4e83a98bf.png">

**Comment by JiahaoYao** [comment]
See http://weinbe58.github.io/QuSpin/generated/quspin.tools.evolution.evolve.html
Times can be an array!

**Reply by Yuanningss** [reply]
Sorry, I did not quite get it.
Currently I have the hamiltonian defined in this fashion:
H = hamiltonian(static , [] , basis=basis , dtype=np.float64 ,**no_checks)
and I have define a initial state psi_0, which is an array of 20 entries. 
I define time as t = np.linspace(t_0 , t_f , num) 
I time evolve the state according to this fashion: psi_t = H.evolve ( psi_0 , t [0] , t) and since I am only interested in the state at t_f, I extract psi_t [ -1 ] hope it only returns the final state. However, the size of the psi_t [ -1 ] is the number of time steps not the actual size of the state, which is 20. 

The question is: how to return the only final state while including the time steps during the simulation ?

**Comment by mgbukov** [comment]
`hamiltonian.evolve` is a wrapper over scipy's adaptive Runge-Kutta; this means that the solver adaptively selects the intermediate time steps, but only returns the solution at the steps specified in `times`.

**Comment by weinbe58** [comment]
> if I want to calculate the state at time t, but with 1000 steps in between, how should I do ?

Are you asking if you can set the time step size of the ODE solver to be (t_f-t_i)/1000?

**Comment by Yuanningss** [comment]
Yes !

**Reply by weinbe58** [reply]
Unfortunately I don't think this is possible. As far as I can the solvers allow you to set a maximum step size but not a minimum. If you are worried about the solver taking too long you can increase `atol` and `rtol` to decrease the number of steps done to evolve the states.
